xbmc/cores/RetroPlayer/rendering/VideoShaders/windows/VideoShaderDX.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ *      Copyright (C) 2017 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this Program; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "VideoShaderDX.h"
+
+#include "Application.h"
+#include "cores/RetroPlayer/rendering/VideoShaders/VideoShaderUtils.h"
+#include "cores/RetroPlayer/rendering/VideoShaders/windows/VideoShaderUtilsDX.h"
+#include "utils/log.h"
+#include "utils/URIUtils.h"
+#include "windowing/WindowingFactory.h"
+
+using namespace KODI;
+using namespace SHADER;
+
+CVideoShaderDX::~CVideoShaderDX()
+{
+  SAFE_RELEASE(m_pInputBuffer);
+}
+
+bool CVideoShaderDX::Create(const std::string& shaderSource, const std::string& shaderPath, ShaderParameters shaderParameters,
+  IShaderSampler* sampler, IShaderLuts luts, float2 viewPortSize, unsigned frameCountMod)
+{
+  m_shaderSource = shaderSource;
+  m_shaderPath = shaderPath;
+  m_shaderParameters = shaderParameters;
+  m_pSampler = reinterpret_cast<ID3D11SamplerState*>(sampler);
+  m_luts = luts;
+  m_viewportSize = viewPortSize;
+  m_frameCountMod = frameCountMod;
+
+  DefinesMap defines;
+
+  defines["HLSL_4"] = "";  // using Shader Model 4
+  defines["HLSL_FX"] = "";  // and the FX11 framework
+
+  // We implement runtime shader parameters ("#pragma parameter")
+  // NOTICE: Runtime shader parameters allow convenient experimentation with real-time
+  //         feedback, as well as override-ability by presets, but sometimes they are
+  //         much slower because they prevent static evaluation of a lot of math.
+  //         Disabling them drastically speeds up shaders that use them heavily.
+  defines["PARAMETER_UNIFORM"] = "";
+
+  m_effect.AddIncludePath(URIUtils::GetBasePath(m_shaderPath));
+
+  if (!m_effect.Create(shaderSource, &defines))
+  {
+    CLog::Log(LOGERROR, "%s: failed to load video shader: %s", __FUNCTION__, shaderPath.c_str());
+    return false;
+  }
+
+  return true;
+}
+
+void CVideoShaderDX::Render(IShaderTexture* source, IShaderTexture* target)
+{
+  auto* sourceDX = static_cast<CShaderTextureCD3D*>(source);
+  auto* targetDX = static_cast<CShaderTextureCD3D*>(target);
+  // TODO: Doesn't work. Investigate calling this in Execute or binding the SRV first
+  //g_Windowing.Get3D11Context()->PSSetSamplers(2, 1, &m_pSampler);
+  SetShaderParameters( *sourceDX->GetPointer() );
+  Execute({ targetDX->GetPointer() }, 4);
```

Comment:
please elaborate on this. What is Execute and what does your comment mean?

Reply:
The shader spec I'm implementing allows for specifying either linear, or nearest sampling for each pass. This line was supposed to set the sampler state to the appropriate ones we've read from the shader preset file with `PSSetSamplers` right before rendering each pass/shader, but it didn't work.

This wasn't a _huge_ problem as these samplers were often hardcoded in the shaders after we started porting them from Cg to HLSL/FX, instead of relying on the implementation reading the shader preset files. Ideally it'd work though, as some shaders files are re-used by different presets with different sampling.

Anyway, I think `Execute` refered to `CWinShader::Execute`, you may as well disregard that because:
Upon further inspection with the Graphics Debugger (_after_ making the comment) it became apparent that the reason this didn't work was because the FX11 framework was applying its _own_ samplers right before rendering, thereby overwriting whatever sampler I was setting here. I couldn't find a way to set the sampler with FX11 or overwrite, so I left it at that as it was good enough for now.